core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -371,6 +371,13 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       throw new SparkException("An application name must be set in your configuration")
     }
 
+    // Thread name has been set to "Driver" if user code ran by AM on a YARN cluster
```

Comment:
Two things here:

- Should also check for `yarn-standalone` since the code still accepts that as an alias for `yarn-cluster`.
- I think the check for `Driver` is too brittle. I think checking for `spark.yarn.app.id` in `SparkConf` is slightly better. Both are not 100% fool-proof since users can muck with them, but at least the latter is an internal API (i.e. the YARN code depends on that property to exist for other things).